File: sdn-controller/test/property/test_properties.ml
(* Property-Based Tests for the SDN Controller
   =============================================
   These tests use QCheck to GENERATE random network topologies and
   verify that INVARIANTS hold for ALL of them — not just the specific
   topologies we thought to test manually.

   WHY PROPERTY-BASED TESTING (for interview):
   Unit tests check: "Does this specific input produce this specific output?"
   Property tests check: "Does this PROPERTY hold for ALL possible inputs?"

   Example:
   - Unit test: "Path from A to C through B costs 150μs" ← one case
   - Property: "For ANY topology, the computed shortest path cost is
     always <= the cost of any other valid path" ← ALL cases

   Property-based testing finds edge cases that humans miss:
   - What happens with a single-node graph?
   - What if all link latencies are 0?
   - What if the graph is fully connected?
   - What if there are 100 nodes?

   Jane Street contributes heavily to the OCaml testing ecosystem
   and uses property-based testing extensively. Showing you test
   this way signals that you think about correctness rigorously.

   KEY PROPERTIES WE TEST:
   1. Path existence: if two nodes are connected, a path exists
   2. Path optimality: Dijkstra always finds the minimum-cost path
   3. Symmetry: path cost A→B equals path cost B→A (undirected graph)
   4. Triangle inequality: direct path is never worse than going through C
   5. Failover: for any single link failure, if the graph was 2-connected,
      an alternate path exists
   6. No routing loops: the path never visits a node twice
*)

open Topology.Network_graph
open Pathcomp.Path_engine

(* ============================================================
   Random Generators for Network Topologies

   QCheck needs to know how to generate random inputs.
   We define generators for devices, links, and full topologies.
   ============================================================ *)

(** Generate a random device ID like "node-0", "node-1", etc. *)
let _gen_device_id =
  QCheck.Gen.(map (fun i -> Printf.sprintf "node-%d" i) (int_range 0 50))

(** Generate a random device with a given ID. *)
let make_random_device id =
  {
    id;
    name = id;
    role = CoreRouter;
    site = "test-site";
    loopback = Printf.sprintf "10.0.0.%d" (Hashtbl.hash id mod 256);
    vendor = "cisco";
    asn = 65001;
  }

(** Generate a random link between two nodes. *)
let make_random_link ~id ~latency =
  {
    id;
    src_interface = "eth0";
    dst_interface = "eth0";
    bandwidth_mbps = 10000;
    state = Up;
    metrics = {
      latency_us = latency;
      jitter_us = latency *. 0.1;
      utilization = 0.1;
      packet_loss = 0.0;
      last_updated = 0.0;
    };
    ospf_cost = max 1 (int_of_float latency);
  }

(** Generate a random connected topology with n nodes.

    We ensure connectivity by creating a spanning tree first
    (connect node i to node i-1), then adding random extra edges.
    This guarantees the graph is connected, which is important
    for testing path existence properties. *)
let gen_connected_topology =
  QCheck.Gen.(
    let* n = int_range 2 15 in  (* 2 to 15 nodes *)
    let* extra_edges = int_range 0 (n * 2) in

    let node_ids = List.init n (fun i -> Printf.sprintf "node-%d" i) in

    (* Build graph with all nodes *)
    let g = List.fold_left (fun g id ->
      add_device g (make_random_device id)
    ) (empty ()) node_ids in

    (* Create spanning tree: connect i to i-1 *)
    let* latencies = list_size (return (n - 1))
      (float_range 1.0 10000.0) in
    let g = List.fold_left2 (fun g i lat ->
      let src = Printf.sprintf "node-%d" i in
      let dst = Printf.sprintf "node-%d" (i - 1) in
      let link_id = Printf.sprintf "%s--%s" src dst in
      add_link g ~src ~dst (make_random_link ~id:link_id ~latency:lat)
    ) g (List.init (n - 1) (fun i -> i + 1)) latencies in

    (* Add random extra edges for redundancy *)
    let* extra = list_size (return extra_edges) (
      let* src_idx = int_range 0 (n - 1) in
      let* dst_idx = int_range 0 (n - 1) in
      let* lat = float_range 1.0 10000.0 in
      return (src_idx, dst_idx, lat)
    ) in
    let g = List.fold_left (fun g (si, di, lat) ->
      if si = di then g  (* No self-loops *)
      else
        let src = Printf.sprintf "node-%d" si in
        let dst = Printf.sprintf "node-%d" di in
        let link_id = Printf.sprintf "%s--%s-extra" src dst in
        (* Only add if link doesn't already exist *)
        match find_link g ~src ~dst with
        | Some _ -> g
        | None ->
          (try add_link g ~src ~dst (make_random_link ~id:link_id ~latency:lat)
           with _ -> g)
    ) g extra in

    return (g, node_ids)
  )

let arbitrary_topology =
  QCheck.make
    ~print:(fun (g, _ids) ->
      Printf.sprintf "Topology(%d nodes, %d links)"
        (List.length (all_devices g))
        (List.length (all_links g)))
    gen_connected_topology

(* ============================================================
   Property 1: Path Existence
   "In a connected graph, a path exists between any two nodes."

   This is the most fundamental property. If Dijkstra can't find
   a path in a connected graph, something is deeply wrong.
   ============================================================ *)

let prop_path_exists_in_connected_graph =
  QCheck.Test.make ~count:200 ~name:"path exists in connected graph"
    arbitrary_topology
    (fun (graph, node_ids) ->
      let n = List.length node_ids in
      if n < 2 then true
      else
        let src = List.nth node_ids 0 in
        let dst = List.nth node_ids (n - 1) in
        match shortest_path graph ~src ~dst
                ~weight:(fun l -> l.metrics.latency_us) with
        | Some _ -> true
        | None -> false)

(* ============================================================
   Property 2: Path Optimality
   "The shortest path cost is <= the cost of any other valid path."

   We verify this by checking that the path from A→B→C is never
   cheaper than the direct shortest path from A→C.
   ============================================================ *)

let prop_shortest_path_is_optimal =
  QCheck.Test.make ~count:200 ~name:"shortest path is optimal"
    arbitrary_topology
    (fun (graph, node_ids) ->
      let n = List.length node_ids in
      if n < 3 then true
      else
        let src = List.nth node_ids 0 in
        let dst = List.nth node_ids (n - 1) in
        let weight link = link.metrics.latency_us in

        match shortest_path graph ~src ~dst ~weight with
        | None -> true  (* No path, nothing to check *)
        | Some (_path, optimal_cost) ->
          (* Check through a random intermediate node *)
          let mid = List.nth node_ids (n / 2) in
          let via_mid_cost =
            match shortest_path graph ~src ~dst:mid ~weight,
                  shortest_path graph ~src:mid ~dst ~weight with
            | Some (_, c1), Some (_, c2) -> Some (c1 +. c2)
            | _ -> None
          in
          match via_mid_cost with
          | None -> true  (* Mid not reachable, skip *)
          | Some indirect ->
            (* Optimal cost should be <= going through mid *)
            optimal_cost <= indirect +. 0.001  (* small epsilon for float *)
    )

(* ============================================================
   Property 3: Symmetry
   "In an undirected graph, cost(A→B) = cost(B→A)."

   Our graph is bidirectional (we add links in both directions),
   so this should always hold.
   ============================================================ *)

let prop_path_symmetry =
  QCheck.Test.make ~count:200 ~name:"path cost is symmetric"
    arbitrary_topology
    (fun (graph, node_ids) ->
      let n = List.length node_ids in
      if n < 2 then true
      else
        let a = List.nth node_ids 0 in
        let b = List.nth node_ids (n - 1) in
        let weight link = link.metrics.latency_us in

        match shortest_path graph ~src:a ~dst:b ~weight,
              shortest_path graph ~src:b ~dst:a ~weight with
        | Some (_, cost_ab), Some (_, cost_ba) ->
          Float.abs (cost_ab -. cost_ba) < 0.001
        | None, None -> true   (* Both unreachable *)
        | _ -> false            (* One reachable, other not — broken *)
    )

(* ============================================================
   Property 4: No Routing Loops
   "A shortest path never visits the same node twice."

   A routing loop would mean traffic circles forever.
   In a trading network, this would be catastrophic.
   ============================================================ *)

let prop_no_routing_loops =
  QCheck.Test.make ~count:200 ~name:"no routing loops in path"
    arbitrary_topology
    (fun (graph, node_ids) ->
      let n = List.length node_ids in
      if n < 2 then true
      else
        let src = List.nth node_ids 0 in
        let dst = List.nth node_ids (n - 1) in
        match shortest_path graph ~src ~dst
                ~weight:(fun l -> l.metrics.latency_us) with
        | None -> true
        | Some (path, _) ->
          let nodes = List.map fst path in
          let unique = List.sort_uniq String.compare nodes in
          List.length nodes = List.length unique)

(* ============================================================
   Property 5: Failover Existence
   "If the graph has redundant paths (is 2-edge-connected),
    removing any single link still leaves a path."

   We test this on topologies with enough extra edges to likely
   be 2-connected. This is the key property for trading networks:
   any single failure should not partition the network.
   ============================================================ *)

let prop_single_failure_survivable =
  QCheck.Test.make ~count:100 ~name:"survives single link failure"
    arbitrary_topology
    (fun (graph, node_ids) ->
      let n = List.length node_ids in
      if n < 3 then true  (* Too small to test meaningfully *)
      else
        let src = List.nth node_ids 0 in
        let dst = List.nth node_ids (n - 1) in
        let links = all_links graph in
        (* Weaker but correct property:
           If removing a link disconnects src from dst, then there was
           only one path. We check that the FAILOVER path (if it exists)
           is still valid — i.e., the controller doesn't crash or loop.
           We also verify: if a path existed before AND after failure,
           the post-failure cost >= pre-failure cost. *)
        let _original_path = shortest_path graph ~src ~dst
            ~weight:(fun l -> l.metrics.latency_us) in
        List.for_all (fun (lsrc, ldst, _link) ->
          let g' = update_link_state graph ~src:lsrc ~dst:ldst Down in
          match shortest_path g' ~src ~dst
                  ~weight:(fun l -> l.metrics.latency_us) with
          | Some (path, _cost) ->
            (* If a path exists after failure, it must have no loops *)
            let nodes = List.map fst path in
            let unique = List.sort_uniq String.compare nodes in
            List.length nodes = List.length unique
          | None -> true  (* Disconnected is OK — just means no redundancy *)
        ) links)

(* ============================================================
   Property 6: Metric Update Consistency
   "Updating a link's metrics changes the path cost appropriately."

   If we increase a link's latency, the shortest path through it
   should cost more (or the algorithm should find a different path).
   ============================================================ *)

let prop_metric_update_affects_cost =
  QCheck.Test.make ~count:200 ~name:"metric updates affect path cost"
    arbitrary_topology
    (fun (graph, node_ids) ->
      let n = List.length node_ids in
      if n < 2 then true
      else
        let src = List.nth node_ids 0 in
        let dst = List.nth node_ids (n - 1) in
        let weight link = link.metrics.latency_us in

        match shortest_path graph ~src ~dst ~weight with
        | None -> true
        | Some (path, cost_before) ->
          if List.length path = 0 then true
          else
            (* Increase latency on the first link in the path *)
            let (first_node, first_link) = List.hd path in
            let new_metrics = { first_link.metrics with
              latency_us = first_link.metrics.latency_us +. 50000.0
            } in
            let g' = update_metrics graph ~src ~dst:first_node new_metrics in
            match shortest_path g' ~src ~dst ~weight with
            | None -> true  (* Extreme case — still valid *)
            | Some (_, cost_after) ->
              (* Cost should either increase or algorithm finds new path *)
              cost_after >= cost_before -. 0.001)

(* ============================================================
   Property 7: JSON Roundtrip
   "Serializing and deserializing a graph preserves all data."

   This is critical for the REST API — if JSON roundtripping
   loses information, the dashboard shows wrong data.
   ============================================================ *)

let prop_json_roundtrip =
  QCheck.Test.make ~count:100 ~name:"JSON roundtrip preserves graph"
    arbitrary_topology
    (fun (graph, _node_ids) ->
      let json = to_yojson graph in
      match of_yojson json with
      | Ok graph2 ->
        let d1 = List.length (all_devices graph) in
        let d2 = List.length (all_devices graph2) in
        let l1 = List.length (all_links graph) in
        let l2 = List.length (all_links graph2) in
        d1 = d2 && l1 = l2
      | Error _ -> false)

(* ============================================================
   Property 8: Traffic Class Ordering
   "Market data routing always produces latency <= trading routing
    for the same src/dst pair."

   Market data weight is pure latency. Trading weight adds jitter
   penalty. So market data path should always be <= trading path
   in terms of raw latency.
   ============================================================ *)

let prop_market_data_fastest =
  QCheck.Test.make ~count:200 ~name:"market data path has lowest latency"
    arbitrary_topology
    (fun (graph, node_ids) ->
      let n = List.length node_ids in
      if n < 2 then true
      else
        let src = List.nth node_ids 0 in
        let dst = List.nth node_ids (n - 1) in
        let md = compute_path graph ~src ~dst
          ~traffic_class:MarketData ~constraints:[] in
        let tr = compute_path graph ~src ~dst
          ~traffic_class:Trading ~constraints:[] in
        match md, tr with
        | Some md_result, Some tr_result ->
          md_result.total_latency_us <= tr_result.total_latency_us +. 0.001
        | None, None -> true
        | Some _, None -> true  (* MD found path but trading didn't — fine *)
        | None, Some _ -> false (* Trading found path but MD didn't — bug *)
    )

(* ============================================================
   Test Runner
   ============================================================ *)

let () =
  let suite = List.map QCheck_alcotest.to_alcotest [
    prop_path_exists_in_connected_graph;
    prop_shortest_path_is_optimal;
    prop_path_symmetry;
    prop_no_routing_loops;
    prop_single_failure_survivable;
    prop_metric_update_affects_cost;
    prop_json_roundtrip;
    prop_market_data_fastest;
  ] in
  Alcotest.run "SDN Controller Properties" [
    "properties", suite;
  ]

File: sdn-controller/tests/test_properties.py
"""Property-based tests for the SDN controller.

Ported from the original OCaml QCheck suite (test_properties.ml). Instead of
checking specific inputs, these GENERATE random network topologies and verify
that INVARIANTS hold across all of them.

The original used QCheck; here we use a small self-contained seeded generator
(Python's ``random``) so the suite has no extra dependencies and reproduces the
"8 properties across 1,600 random topologies" claim exactly (8 × 200 = 1,600).
Seeds are fixed so failures are reproducible.

Properties verified:
  1. Path existence — connected graph always has a path
  2. Path optimality — Dijkstra finds the minimum-cost path
  3. Symmetry — cost(A→B) == cost(B→A) in an undirected graph
  4. No routing loops — a path never visits a node twice
  5. Failover — removing any single link never produces a looped path
  6. Metric-update consistency — raising a link's latency never lowers path cost
  7. JSON roundtrip — serialize/deserialize preserves the graph
  8. Traffic-class ordering — market-data latency <= trading latency
"""

import random

from tradenet_sdn.network_graph import (
    Device,
    DeviceRole,
    Link,
    LinkMetrics,
    LinkState,
    NetworkGraph,
)
from tradenet_sdn.path_engine import TrafficClass, compute_path

COUNT = 200  # iterations per property (8 × 200 = 1,600 random topologies)
EPS = 0.001  # float comparison epsilon


def latency_weight(link):
    return link.metrics.latency_us


# ============================================================
# Random generators
# ============================================================


def make_random_device(node_id):
    try:
        octet = int(node_id.split("-")[1]) % 256
    except (IndexError, ValueError):
        octet = 0
    return Device(id=node_id, name=node_id, role=DeviceRole.CORE_ROUTER,
                  site="test-site", loopback=f"10.0.0.{octet}", vendor="cisco", asn=65001)


def make_random_link(link_id, latency):
    return Link(
        id=link_id,
        src_interface="eth0",
        dst_interface="eth0",
        bandwidth_mbps=10000,
        state=LinkState.UP,
        metrics=LinkMetrics(latency_us=latency, jitter_us=latency * 0.1,
                            utilization=0.1, packet_loss=0.0, last_updated=0.0),
        ospf_cost=max(1, int(latency)),
    )


def gen_connected_topology(rng):
    """Generate a random connected topology with 2-15 nodes.

    Connectivity is guaranteed by first building a spanning tree (connect node i
    to node i-1), then adding random extra edges for redundancy. Mirrors the
    original OCaml ``gen_connected_topology``.
    """
    n = rng.randint(2, 15)
    extra_edges = rng.randint(0, n * 2)
    node_ids = [f"node-{i}" for i in range(n)]

    g = NetworkGraph.empty()
    for node_id in node_ids:
        g = g.add_device(make_random_device(node_id))

    # Spanning tree: connect node-i to node-(i-1)
    for i in range(1, n):
        src = f"node-{i}"
        dst = f"node-{i - 1}"
        latency = rng.uniform(1.0, 10000.0)
        g = g.add_link(src, dst, make_random_link(f"{src}--{dst}", latency))

    # Random extra edges for redundancy
    for _ in range(extra_edges):
        si = rng.randint(0, n - 1)
        di = rng.randint(0, n - 1)
        latency = rng.uniform(1.0, 10000.0)
        if si == di:
            continue  # No self-loops
        src = f"node-{si}"
        dst = f"node-{di}"
        if g.find_link(src, dst) is not None:
            continue  # Edge already exists
        try:
            g = g.add_link(src, dst, make_random_link(f"{src}--{dst}-extra", latency))
        except Exception:
            pass

    return g, node_ids


def topologies(base_seed):
    """Yield (iteration, graph, node_ids) for COUNT reproducible random topologies."""
    for i in range(COUNT):
        rng = random.Random(base_seed + i)
        graph, node_ids = gen_connected_topology(rng)
        yield i, graph, node_ids


# ============================================================
# Property 1: Path existence
# ============================================================


def test_path_exists_in_connected_graph():
    for i, graph, node_ids in topologies(1_000_000):
        n = len(node_ids)
        if n < 2:
            continue
        src, dst = node_ids[0], node_ids[n - 1]
        assert graph.shortest_path(src, dst, latency_weight) is not None, \
            f"iteration {i}: no path in connected graph ({n} nodes)"


# ============================================================
# Property 2: Path optimality
# ============================================================


def test_shortest_path_is_optimal():
    for i, graph, node_ids in topologies(2_000_000):
        n = len(node_ids)
        if n < 3:
            continue
        src, dst = node_ids[0], node_ids[n - 1]
        opt = graph.shortest_path(src, dst, latency_weight)
        if opt is None:
            continue
        _path, optimal_cost = opt

        mid = node_ids[n // 2]
        first = graph.shortest_path(src, mid, latency_weight)
        second = graph.shortest_path(mid, dst, latency_weight)
        if first is None or second is None:
            continue  # mid not reachable, skip
        indirect = first[1] + second[1]
        assert optimal_cost <= indirect + EPS, \
            f"iteration {i}: optimal {optimal_cost} > via-mid {indirect}"


# ============================================================
# Property 3: Symmetry
# ============================================================


def test_path_symmetry():
    for i, graph, node_ids in topologies(3_000_000):
        n = len(node_ids)
        if n < 2:
            continue
        a, b = node_ids[0], node_ids[n - 1]
        ab = graph.shortest_path(a, b, latency_weight)
        ba = graph.shortest_path(b, a, latency_weight)
        if ab is not None and ba is not None:
            assert abs(ab[1] - ba[1]) < EPS, \
                f"iteration {i}: cost(a→b)={ab[1]} != cost(b→a)={ba[1]}"
        else:
            assert ab is None and ba is None, \
                f"iteration {i}: one direction reachable, the other not"


# ============================================================
# Property 4: No routing loops
# ============================================================


def test_no_routing_loops():
    for i, graph, node_ids in topologies(4_000_000):
        n = len(node_ids)
        if n < 2:
            continue
        src, dst = node_ids[0], node_ids[n - 1]
        found = graph.shortest_path(src, dst, latency_weight)
        if found is None:
            continue
        path, _cost = found
        nodes = [node for node, _link in path]
        assert len(nodes) == len(set(nodes)), \
            f"iteration {i}: routing loop in path {nodes}"


# ============================================================
# Property 5: Failover existence (no looped failover path)
# ============================================================


def test_single_failure_survivable():
    for i, graph, node_ids in topologies(5_000_000):
        n = len(node_ids)
        if n < 3:
            continue
        src, dst = node_ids[0], node_ids[n - 1]
        for lsrc, ldst, _link in graph.all_links():
            g2 = graph.update_link_state(lsrc, ldst, LinkState.DOWN)
            found = g2.shortest_path(src, dst, latency_weight)
            if found is None:
                continue  # Disconnected is OK — just no redundancy
            path, _cost = found
            nodes = [node for node, _link in path]
            assert len(nodes) == len(set(nodes)), \
                f"iteration {i}: looped failover path after dropping {lsrc}--{ldst}"


# ============================================================
# Property 6: Metric-update consistency
# ============================================================


def test_metric_update_affects_cost():
    for i, graph, node_ids in topologies(6_000_000):
        n = len(node_ids)
        if n < 2:
            continue
        src, dst = node_ids[0], node_ids[n - 1]
        found = graph.shortest_path(src, dst, latency_weight)
        if found is None:
            continue
        path, cost_before = found
        if len(path) == 0:
            continue
        first_node, first_link = path[0]
        new_metrics = LinkMetrics(
            latency_us=first_link.metrics.latency_us + 50000.0,
            jitter_us=first_link.metrics.jitter_us,
            utilization=first_link.metrics.utilization,
            packet_loss=first_link.metrics.packet_loss,
            last_updated=first_link.metrics.last_updated,
        )
        g2 = graph.update_metrics(src, first_node, new_metrics)
        found2 = g2.shortest_path(src, dst, latency_weight)
        if found2 is None:
            continue
        _path2, cost_after = found2
        assert cost_after >= cost_before - EPS, \
            f"iteration {i}: cost dropped after raising latency ({cost_before} -> {cost_after})"


# ============================================================
# Property 7: JSON roundtrip
# ============================================================


def test_json_roundtrip():
    for i, graph, _node_ids in topologies(7_000_000):
        data = graph.to_dict()
        graph2 = NetworkGraph.from_dict(data)
        assert len(graph.all_devices()) == len(graph2.all_devices()), \
            f"iteration {i}: device count changed across roundtrip"
        assert len(graph.all_links()) == len(graph2.all_links()), \
            f"iteration {i}: link count changed across roundtrip"


# ============================================================
# Property 8: Traffic-class ordering
# ============================================================


def test_market_data_fastest():
    for i, graph, node_ids in topologies(8_000_000):
        n = len(node_ids)
        if n < 2:
            continue
        src, dst = node_ids[0], node_ids[n - 1]
        md = compute_path(graph, src, dst, TrafficClass.MARKET_DATA, constraints=[])
        tr = compute_path(graph, src, dst, TrafficClass.TRADING, constraints=[])
        if md is not None and tr is not None:
            assert md.total_latency_us <= tr.total_latency_us + EPS, \
                f"iteration {i}: market-data latency {md.total_latency_us} > trading {tr.total_latency_us}"
        elif md is None and tr is not None:
            raise AssertionError(f"iteration {i}: trading found a path but market data didn't")
        # md found / tr not, or both none: fine
